File: memories/2024/exercise/java/day13/src/main/java/santamarket/model/ShoppingSleigh.java
package santamarket.model;

import java.util.*;

public class ShoppingSleigh {
    private final List<ProductQuantity> items = new ArrayList<>();

    public void addItemQuantity(Product product, double quantity) {
        items.add(new ProductQuantity(product, quantity));
    }

    public List<ProductQuantity> getItems() {
        return new ArrayList<>(items);
    }

    public void handleOffers(Receipt receipt, Map<Product, Offer> offers, SantamarketCatalog catalog) {
        Map<Product, Double> productQuantities = new HashMap<>();
        for (ProductQuantity item : items) {
            Product p = item.getProduct();
            if (productQuantities.containsKey(p)) {
                productQuantities.put(p, productQuantities.get(p) + item.getQuantity());
            } else {
                productQuantities.put(p, item.getQuantity());
            }
        }

        for (Product p : productQuantities.keySet()) {
            double quantity = productQuantities.get(p);
            if (offers.containsKey(p)) {
                Offer offer = offers.get(p);
                double unitPrice = catalog.getUnitPrice(p);
                int quantityAsInt = (int) quantity;
                double discountAmount = 0;

                if (offer.offerType == SpecialOfferType.THREE_FOR_TWO) {
                    if (quantityAsInt > 2) {
                        int numberOfSets = quantityAsInt / 3;
                        discountAmount = numberOfSets * unitPrice;
                    }
                } else if (offer.offerType == SpecialOfferType.TEN_PERCENT_DISCOUNT) {
                    discountAmount = quantity * unitPrice * 0.1;
                }

                if (discountAmount > 0) {
                    receipt.addDiscount(discountAmount);
                }
            }
        }
    }
}

File: memories/2024/exercise/C#/day13/SantaMarket/Model/ShoppingSleigh.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SantaMarket.Model
{
    public class ShoppingSleigh
    {
        private readonly List<ProductQuantity> items = new List<ProductQuantity>();

        public void AddItemQuantity(Product product, double quantity)
        {
            items.Add(new ProductQuantity(product, quantity));
        }

        public IReadOnlyList<ProductQuantity> GetItems()
        {
            return items.AsReadOnly();
        }

        public void HandleOffers(Receipt receipt, Dictionary<Product, Offer> offers, ISantamarketCatalog catalog)
        {
            var productQuantities = new Dictionary<Product, double>();
            foreach (var item in items)
            {
                var p = item.Product;
                if (productQuantities.ContainsKey(p))
                {
                    productQuantities[p] += item.Quantity;
                }
                else
                {
                    productQuantities[p] = item.Quantity;
                }
            }

            foreach (var p in productQuantities.Keys)
            {
                var quantity = productQuantities[p];
                if (offers.ContainsKey(p))
                {
                    var offer = offers[p];
                    var unitPrice = catalog.GetUnitPrice(p);
                    var quantityAsInt = (int)quantity;
                    var discountAmount = 0.0;

                    if (offer.OfferType == SpecialOfferType.ThreeForTwo)
                    {
                        if (quantityAsInt > 2)
                        {
                            var numberOfSets = quantityAsInt / 3;
                            discountAmount = numberOfSets * unitPrice;
                        }
                    }
                    else if (offer.OfferType == SpecialOfferType.TenPercentDiscount)
                    {
                        discountAmount = quantity * unitPrice * 0.1;
                    }

                    if (discountAmount > 0)
                    {
                        receipt.AddDiscount(discountAmount);
                    }
                }
            }
        }
    }
}
